system-addon/content-src/lib/init-store.js

```diff
@@ -76,6 +76,29 @@ const rehydrationMiddleware = store => next => action => {
   return next(action);
 };
 
+// This middleware queues up all the SAVE_SESSION_PERF_DATA actions until it
+// receives the MERGE_STORE_ACTION, which is guaranteed to come from chrome.
```

Comment:
There's plenty of messages from main (I think we should be consistent with just calling chrome "main"), that would allow the queue to be emptied. @k88hudson Would it be safe to treat any `!au.isSendToMain(action)` as coming from main? I suppose some content could dispatch to itself, so I guess not. Although we can just as easily add an `au.isFromMain(action)`.

Then I would see this early queue behaving like:
```js
… => {
  if (store._receivedFromMain || action.type === at.NEW_TAB_STATE_REQUEST) {
    next(action);
  } else if (au.isFromMain(action)) {
    // empty the queue
  } else {
    // add to queue
  }
}
```